src/Analyzers/MSTest.Analyzers/DataRowShouldBeValidAnalyzer.cs

```diff
@@ -0,0 +1,214 @@
+        if (attribute.ConstructorArguments.Length == 0)
+        {
+            if (!lastMethodParameterIsArray)
+            {
+                context.ReportDiagnostic(syntax.CreateDiagnostic(
+                    ArgumentCountMismatchRule,
+                    attribute.ConstructorArguments.Length,
+                    methodSymbol.Parameters.Length));
+            }
+
+            return;
+        }
+
+        // DataRow constructors have either zero or one argument(s). If we get here, we are
+        // on the one argument case. Check if we match either of the array argument constructors
+        // and expand the array argument if we do.
+        ImmutableArray<TypedConstant> constructorArguments = attribute.ConstructorArguments;
+        if (constructorArguments[0].Kind is TypedConstantKind.Array && !constructorArguments[0].IsNull)
+        {
+            constructorArguments = constructorArguments[0].Values;
+        }
+
+        if (IsArgumentCountMismatch(constructorArguments.Length, methodSymbol.Parameters.Length, lastMethodParameterIsArray))
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentCountMismatchRule,
+                constructorArguments.Length,
+                methodSymbol.Parameters.Length));
+            return;
+        }
+
+        // Check constructor argument types match method parameter types.
+        List<(int ConstructorArgumentIndex, int MethodParameterIndex)> typeMismatchIndices = new();
+        for (int i = 0, j = 0; i < constructorArguments.Length; ++i, ++j)
+        {
+            if (constructorArguments[i].IsNull)
+            {
+                continue;
+            }
+
+            ITypeSymbol? argumentType = constructorArguments[i].Type;
+            ITypeSymbol paramType = (lastMethodParameterIsArray && j >= methodSymbol.Parameters.Length - 1)
+                ? ((IArrayTypeSymbol)lastMethodParameter.Type).ElementType
+                : methodSymbol.Parameters[j].Type;
+
+            if (argumentType is not null && !argumentType.IsAssignableTo(paramType, context.Compilation))
+            {
+                typeMismatchIndices.Add((i, j));
+            }
+        }
+
+        // Report diagnostics if there's any type mismatch.
+        if (typeMismatchIndices.Count > 0)
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentTypeMismatchRule,
+                FormatTypeMismatchIndexList(typeMismatchIndices)));
+        }
+    }
+
+    private static bool IsArgumentCountMismatch(int constructorArgumentsLength, int methodParametersLength, bool lastMethodParameterIsArray)
+    {
+        // 1. If last method parameter is not an array the lengths must be the same.
+        // 2. If last method parameter is an array the argument count check is relaxed and we only
+        //    need to make sure we don't have less constructor arguments than actual method paramters.
```

Comment:
I am surprised by the explanation. I would expect the following cases:

1. last parameter is params, then we need arguments count >= parameters count - 1
2. single parameter as array, arguments count >= 0
3. otherwise both lengths should match.

Reply:
Cases 1 and 3 are already handled by this condition. Case 2 is directly handled by the algorithm in the sense that this is a supported scenario. We iterate through each constructor argument and compare against method parameter type. If there's a type mismatch then we report diagnostics, but the argument count rule doesn't apply here.

Reply:
> I don't see a test case with something like `int a, object b` and `[DataRow(1, 2, 3)]`.

There was something similar with 2 int args:

![image](https://github.com/microsoft/testfx/assets/6165373/dbd1aa9d-e894-4900-a0b9-6575cb366776)

I've added the test case you requested too though in case the 2nd param being an object is relevant:

![image](https://github.com/microsoft/testfx/assets/6165373/6c06b582-32ec-4a5d-a6ff-ae78dec55c07)

Reply:
Sorry I gave a wrong example, I mean `int i, object[] o` the test no longer report count mismatch.

What you want is not checking last parameter is array but last parameter is params.

Reply:
This is now fixed as well, thanks !